torch/_dynamo/guards.py

```diff
@@ -512,18 +519,33 @@ def tensor_ref_as_str(tensor_ref, id_to_name_map):
             return f"{id_to_name_map[tensor_ref.ref_id]}.{tensor_ref.kind}()[{tensor_ref.idx}]"
         return f"{id_to_name_map[tensor_ref.ref_id]}.{tensor_ref.kind}()"
 
-    def __init__(self, expr_to_tensor_ref, id_to_name_map):
+    def __init__(
+        self, expr_to_tensor_ref, id_to_name_map, shape_env, intermediary_symbols
+    ):
         super().__init__()
         self.expr_to_tensor_ref = expr_to_tensor_ref
         self.id_to_name_map = id_to_name_map
+        self.shape_env = shape_env
+        self.intermediary_symbols = intermediary_symbols
 
     def _print_Symbol(self, expr) -> str:
         assert isinstance(expr, sympy.core.symbol.Symbol)
         if expr == 0:
             return "0"
         if expr == 1:
             return "1"
-        assert expr in self.expr_to_tensor_ref, f"Unknown expression {expr}"
+        if expr not in self.expr_to_tensor_ref:
+            # Please keep these 2 lines here for debugging
+            # if expr not in self.intermediary_symbols:
+            # log.warning(f"DROPPING GUARD SYMBOL: {expr}")
+            # This is an intermediary symbol with no tensor association, skip it
+            # If we did not make the symbol, it came from something dynamo does not know about
+            # So either: (A) skipping the guard is safe because something else (like module id check)
+            # Cover it. This happens for things like channel in/out on conv2d,
+            # Where changing those will break
+            # other guards - or (B) it is not and we made a mistake, hence the warning above.
```

Comment:
This doesn't look right. The only symbols that should show up in expressions anywhere should correspond to inputs to the dynamo trace. It would be better to investigate the case that triggers this (channel in/out on conv2d) and root cause further.

Reply:
We know the case - its because we did not trace all modules at the same level of detail as tensors. The point is that a hypothetical symbolic integer (sN) is coming from an object dynamo saw, but is not a tensor, and therefore does not track details of it. These objects are safe to not guard on via symbolic shapes, because we guard on their identity. We will have guards for it in the form of an id check and type check, but we do not need these symbolic-shape guards as dynamo is not capturing the object at this level, and does not need to guard it (the existing dynamo specialization is stricter and more likely to invalidate than this).

Reply:
If this is the case, why are we creating the guard in the first place just so we can ignore it later?

Reply:
Shape env doesn't know that we have a stricter guard in place - it's just reflecting what is being traced. It sees an integer, (say, X in a channel value on a conv2d), and makes a guard like sN == 256. Internally to conv2d, it may do some comparison, and then a new guard is born, fake ex, sN > sK ^ 2. None of these matter to us, because we are specializing on the entire conv2d identity and type (idx 12345, type Conv2D). If these changes, we break the guard and recompile. Specifically, these values are *not* on tensors and have nothing to do with dynamic shapes, we are just getting them in shape_env as a side effect of tracing. 

Specializing on the shape relationship of these number does not make for any kind of stricter guarding than id/type. THE ONLY edge case we are failing to protect is if someone manually changes the values of these integers somewhere, but iirc dynamo already does not handle this, so we are not introducing any new issues. 

Happy to be wrong here though, so please lmk if I missed anything.

Reply:
Can you provide an example of where these guards are coming from?  I still don't fully understand.  I'm having trouble reasoning about why it is safe to ignore some guards.

Reply:
Discussed offline (huge thanks to @ezyang and @Chillee) - deepcopy of nn module is calling fake_mode.from_tensor and making symbolic intermediary shapes. We dont need this branch.